E-Freezer/freezer_sched_class.md

```diff
@@ -0,0 +1,704 @@
+time. At a high level, Linux context switches from process to process, letting
+the CPU perform work on behalf of each one in turn. This switching occurs
+quickly enough to create the illusion that all processes are running
+simultaneously. **The scheduler is in charge of coordinating all of this
+switching.**
+
+### Where Does the Scheduler Fit?
+
+You can find most scheduler-related code under `kernel/sched`. Now, the
+scheduler has a distinct and non-trivial job. The rest of the kernel doesn't
+know or care how the scheduler performs its magic, as long as it can be called
+upon to schedule tasks. So, to hide the complexity of the scheduler, it is
+invoked with a simple and well-defined API. The scheduler - from the perspective
+of the rest of the kernel - has two main responsibilities:
+
+* **Responsibility I**: Provide an interface to halt the currently running
+process and switch to a new one. To do so, the scheduler must pick the next
+process to run, which is a nontrivial problem.
+
+* **Responsibility II**: Indicate to the rest of the OS when a new process
+should be run.
+
+### Responsibility I: Switching to the Next Process
+
+To fulfill its first responsibility, the scheduler must somehow keep track of
+all the running processes.
+
+### The Runqueue
+
+Here's the first over-simplification: you can think of the scheduler as a system
+that maintains a simple queue of processes in the form of a linked list. The
+process at the head of the queue is allowed to run for some "time slice" - say,
+10 milliseconds. After this time slice expires, the process is moved to the back
+of the queue, and the next process gets to run on the CPU for the same time
+slice. When a running process is forcibly stopped and taken off the CPU in this
+way, we say that it has been **preempted**. The linked list of processes waiting
+to have a go on the CPU is called the runqueue. Each CPU has its own runqueue,
+and a given process may appear on only one CPU's runqueue at a time. Processes
+CAN migrate between various CPUs' runqueues, but we'll save this discussion for
+later.
+
+<div align='center'>
+    <img src='./final_simple_runqueue.png'/><br/>
+    Figure 1: An over-simplification of the runqueue
+</div>
+
+The scheduler is not really this simple; the runqueue is defined in the kernel
+as `struct rq`, and you can take a peek at its definition
+[here](https://elixir.bootlin.com/linux/v5.10.158/source/kernel/sched/sched.h#L897).
+Spoiler alert: it's not a linked list! To be fair, the explanation that I gave
+above more or less describes the very first Linux runqueue. But over the years,
+the scheduler evolved to incorporate multiple scheduling algorithms. These
+include:
+
+  * Completely Fair Scheduler (CFS)
+  * Real-Time Scheduler
+  * Deadline Scheduler
+
+The modern-day runqueue is no longer a linked list but actually a collection of
+algorithm-specific runqueues corresponding to the list above. Indeed,

```

Comment:
"For example, the Completely Fair Scheduler (CFS), which is the default scheduler in modern Linux kernels, uses a red-black tree data structure to keep track of processes, with each process assigned a "virtual runtime" that determines its priority in the scheduling queue. The scheduler then selects the process with the lowest virtual runtime to run next, ensuring that each process gets a fair share of CPU time over the long term."